Sources/TuistCore/Cache/CacheDirectoriesProvider.swift

```diff
@@ -0,0 +1,46 @@
+import Foundation
+import TSCBasic
+import TuistGraph
+import TuistSupport
+
+public protocol CacheDirectoriesProviding: AnyObject {
+    /// Returns the cache directory
+    var cacheDirectory: AbsolutePath { get }
+
+    /// Directory where the projects generated for automation are located.
+    var projectsCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where the build artifacts are cached.
+    var buildCacheDirectory: AbsolutePath { get }
+
+    /// Returns the directory where hashes of modules that have been a part of successful test are cached
+    var testsCacheDirectory: AbsolutePath { get }
+}
+
+public class CacheDirectoriesProvider: CacheDirectoriesProviding {
+    public let cacheDirectory: AbsolutePath
+    static let defaultDirectory = AbsolutePath(URL(fileURLWithPath: NSHomeDirectory()).path).appending(component: ".tuist")
+
+    public init(config: Config?) {
+        if let cacheDirectory = config?.cache?.path {
+            self.cacheDirectory = cacheDirectory
+        } else if let envVariable = ProcessInfo.processInfo.environment[Constants.EnvironmentVariables.cacheDirectory] {
+            cacheDirectory = AbsolutePath(envVariable)
+        } else {
+            cacheDirectory = CacheDirectoriesProvider.defaultDirectory.appending(component: "Cache")
+        }
+    }
+
+    public var testsCacheDirectory: AbsolutePath {
+        cacheDirectory.appending(component: "TestsCache")
+    }
+
+    /// Returns the directory where the build artifacts are cached.
```

Comment:
nit: lets keep the docs in the protocol since less places to maintain.
```suggestion
    /// Returns the directory where the build artifacts are cached.
```